### PR TITLE
Make type dispatcher from `cupy.cuda.cub` reusable

### LINKED ISSUE #1183: Remove cupy_common.h

### PR DESCRIPTION
**Pull request by leofang**
~~Blocked by #3386 and #3461.~~

Close #1183.

I refactored a bit the two modules `cupy.cuda.cub` and `cupy.cuda.thrust` so that the C++ type dispatcher from `cupy/cuda/cupy_cub.cu` is moved to `cupy/core/include/cupy/type_dispatcher.cuh` and so can be used in both modules. This has the additional benefit that it ensures both modules see the same definitions of types, in particular the complex numbers. Moreover, on the Python side the "dictionary" between NumPy dtypes and the corresponding C++ constants is moved to `cupy.cuda.common`. The unnecessary code petition is greatly reduced.

### CODE REVIEW

**Comment by leofang** [comment]
I'll resolve the conflicts after the mentioned PRs are merged.

**Comment by leofang** [comment]
After this PR is merged, `cupy_common.h` will no longer be needed, so I removed it in 22a51c0 to close #1183.

**Comment by leofang** [comment]
Jenkins, test this please.

**Comment by leofang** [comment]
(Just wanna run a quick check to confirm.)

**Comment by leofang** [comment]
Power cycle to see if RTD can be re-triggered.

**Comment by leofang** [comment]
Conflicts resolved. 
Jenkins, test this please

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Conflicts resolved. PTAL 🙂

**Comment by leofang** [comment]
Weird, I added a new commit but it's not shown here. Let me power-cycle...

**Comment by leofang** [comment]
Jenkins, test this please

File: cupy/cuda/cupy_thrust.h [hunk 1]
```diff
@@ -8,63 +8,29 @@ namespace cupy {
...
 } // namespace cupy
 
 #else // CUPY_NO_CUDA
 
-#include "../../cupy_backends/cuda/cupy_common.h"
 #define THRUST_VERSION 0
 
 namespace cupy {
 
 namespace thrust {
```
Comment (takagi) [inline cupy/cuda/cupy_thrust.h:25]: Now I think these namespaces do not make much sense. How about removing them like `cupy_cub.h`?

File: cupy/cuda/cupy_thrust.h [hunk 2]
```diff
@@ -8,63 +8,29 @@ namespace cupy {
...
-}
-
-template <typename T>
-void _argsort(size_t *, void *, void *, const std::vector<ptrdiff_t>&, intptr_t,
-              void *) {
-    return;
-}
-
-void _sort_fp16(void *, size_t *, const std::vector<ptrdiff_t>&, intptr_t, void *) {
+void thrust_sort(int, void *, size_t *, const std::vector<ptrdiff_t>&, intptr_t, void *) {
```
Comment (takagi) [inline cupy/cuda/cupy_thrust.h:27]:
```suggestion
void thrust_sort(...) {
```

File: cupy/cuda/cupy_thrust.h [hunk 3]
```diff
@@ -8,63 +8,29 @@ namespace cupy {
...
 }
 
-void _lexsort_fp16(size_t *, void *, size_t, size_t, intptr_t, void *) {
+void thrust_lexsort(int, size_t *, void *, size_t, size_t, intptr_t, void *) {
```
Comment (takagi) [inline cupy/cuda/cupy_thrust.h:30]:
```suggestion
void thrust_lexsort(...) {
```

File: cupy/cuda/cupy_thrust.h [hunk 4]
```diff
@@ -8,63 +8,29 @@ namespace cupy {
...
 }
 
-void _argsort_fp16(size_t *, void *, void *, const std::vector<ptrdiff_t>&, intptr_t,
-                   void *) {
+void thrust_argsort(int, size_t *, void *, void *, const std::vector<ptrdiff_t>&, intptr_t, void *) {
```
Comment (takagi) [inline cupy/cuda/cupy_thrust.h:33]:
```suggestion
void thrust_argsort(...) {
```

**Comment by takagi** [comment]
LGTM except for just minor comments!

File: cupy/cuda/cupy_thrust.h (hunk 1 above)
Comment (leofang) [reply cupy/cuda/cupy_thrust.h:25]: Yeah sounds good. Do you remember why we needed the namespaces?

**Comment by leofang** [comment]
@takagi I removed the namespace!

**Comment by takagi** [comment]
pfnCI, test this please.

File: cupy/cuda/cupy_thrust.h (hunk 1 above)
Comment (takagi) [reply cupy/cuda/cupy_thrust.h]: I initially introduced the namespace but was not necessary. It makes no harm to remove it.

**Comment by takagi** [comment]
LGTM!

**Comment by leofang** [comment]
Thanks!

**Pull request merged**